File: samples/start-app/js-node-express/use-cases/app-activation/callback.js
let doAppCallback = function({
    strings,
    crypto,
    querystring,
    httpsClient,
    Codec,
    tokenDb,
    LogicError,
}) {
    return async function appCallback ({req, res, next}, randomString, openIdCallback=null) {

        if (!req.query.hasOwnProperty('state') || req.query.state !== randomString) {
            res.render('error');
            throw new LogicError('Invalid state');
        }

        const codeIdentifier = strings.bin2hex(crypto.randomBytes(30));
        const codeChallenge = crypto.createHash('sha256').update(codeIdentifier + process.env.CLIENT_SECRET).digest('hex');

        const data = querystring.stringify({
            client_id: process.env.CLIENT_ID,
            code: req.query.code,
            grant_type: "authorization_code",
            code_identifier: codeIdentifier,
            code_challenge: codeChallenge
        });

        const options = {
            path: "/connect/apps/v1/oauth2/token",
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            }
        };

        const response = await httpsClient.request(options, data);

        const access_token = response.access_token;

        if (access_token !== undefined) {
            tokenDb.upsert({access_token});

            if (!await tokenDb.hasToken()) {
                res.render('no_access_token');
                throw new LogicError('Missing access token in database');
            } else {
                if (process.env.OPENID_AUTHENTICATION === "1" && openIdCallback instanceof Function) {
                    const id_token = response.id_token;
                    const user = await openIdCallback(id_token);
                    const encodedCookie = Codec.encoder(user.sub, process.env.SUB_HASH_KEY)
                    res.cookie('sub', encodedCookie.sub);
                    res.cookie('vector', encodedCookie.vector);
                    res.render('access_token', {user: user});
                } else {
                    res.render('access_token');
                }
            }
        } else {
            res.render('no_access_token');
            throw new LogicError('Missing access token in response');
        }
    };
}

module.exports = doAppCallback;

File: samples/start-app/python-fastapi/app/usecase/callback.py
from app.persistence import tokenRepository
from app.model import schemas
from app.dependencies import get_config, build_user_agent
import secrets, requests, hashlib

get_app_token_url = '/connect/apps/v1/oauth2/token'

def callback_usecase(request, db, session):
    session_state = session.headers['oauth2_state']
    state = request.query_params['state'] or ''
    if state == '' or state != session_state:
        raise Exception('Invalide state')

    authorization_code = request.query_params['code'] or ''
    if authorization_code == '':
        raise Exception('Missing authorization code')

    pim_url = session.headers['pim_url']
    if pim_url == '':
        raise Exception('No PIM url in session')

    code_identifier = secrets.token_hex(30)
    code_challenge = hashlib.sha256((code_identifier + get_config('CLIENT_SECRET')).encode('utf-8')).hexdigest()

    access_token_url = pim_url+'%s' % get_app_token_url

    access_token_request_payload = {
        'client_id': get_config('CLIENT_ID'),
        'code_identifier': code_identifier,
        'code_challenge': code_challenge,
        'code': authorization_code,
        'grant_type': 'authorization_code'
    }

    access_token_request_headers = {
        'Content-Type': 'application/x-www-form-urlencoded',
        'User-Agent':  build_user_agent()
    }

    response = requests.post(access_token_url, data=access_token_request_payload, headers=access_token_request_headers)

    token = schemas.TokenCreate(access_token=response.json()['access_token'])

    tokenRepository.create_token(db=db, token=token)
    
    return response
